Vault Specialized Container, Test, and Repo files

mikeoz0 opened the pull request:
This PR includes the initial codebase for the Vault specialized container. I have added a test class and README that includes an example of usage.

Resolves https://github.com/testcontainers/testcontainers-java/issues/331

Commit:
Commit of Vault container, test, and repo files

bsideup commented on .travis.yml:
```diff
@@ -0,0 +1,24 @@
+services:
+  - docker
+
+before_install:
```
these lines are only needed for testing of "the latest and greatest" Docker. I think you can remove the whole "before_install" block

mikeoz0 replied:
done.

bsideup commented on README.md:
```diff
@@ -1,2 +1,52 @@
+
+See [testcontainers.org](https://www.testcontainers.org) for more information about Testcontainers.
+
+<!--[![Build Status](https://travis-ci.org/testcontainers/testcontainers-java.svg?branch=master)](https://travis-ci.org/testcontainers/testcontainers-java)-->
```
this badge is for the main project (testcontainers-java). We should use repo-specific badge here

mikeoz0 replied:
done.

bsideup commented on README.md:
```diff
@@ -1,2 +1,52 @@
+
+## Usage example
+
+Running Vault in your Junit tests is easily done with an @TestRule such as the following:
```
minor: It's just `@Rule`, not `@TestRule` :)

mikeoz0 replied:
done.

bsideup commented on README.md:
```diff
@@ -1,2 +1,52 @@
+public class SomeTest {
+
+    @ClassRule
+    public static VaultContainer vaultContainer = new VaultContainer<>("vault:0.7.0")
```
Maybe we should provide the default constructor with pre-set vault version like we already do for PostgreSQL/MySQL/others?

mikeoz0 replied:
Update made.

bsideup commented on pom.xml:
```diff
@@ -0,0 +1,72 @@
+        <repository>
+            <id>bintray</id>
+            <name>richnorth-maven</name>
+            <url>https://api.bintray.com/maven/richnorth/maven/org.testcontainers.vault/;publish=1</url>
```
@rnorth was it agreed?

mikeoz0 replied:
@rnorth let me know if I need to modify this.

kiview replied:
Maybe it's best to remove this part for the first release an we'll talk about possible artifact management in the future.

bsideup commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,139 @@
+
+    @Override
+    protected void waitUntilContainerStarted() {
+        getWaitStrategy().waitUntilReady(this);
```
you shouldn't call it directly. Use `containerIsStarted` instead

mikeoz0 replied:
done.

bsideup commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    private void addSecrets() {
+        secretsMap.forEach((path, secrets) -> {
```
please run a single exec command inside a container instead of a command per secret, it might affect the performance

mikeoz0 replied:
I have reworked the method to now run a single exec command even if there are multiple secrets or paths.

bsideup commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,139 @@
+        list.add(1, "write");
+        list.add(2, path);
+        String[] array = list.toArray(new String[list.size()]);
+        return array;
```
or just `return new String[] { "vault", "write", path }` :)

mikeoz0 replied:
reworked, thanks

bsideup commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,139 @@
+        setPreloadSecret(true);
+        List<String> list = new ArrayList<>();
+        list.add(firstSecret);
+        Stream.of(remainingSecrets).forEach(s -> list.add(s));
```
please use classic for loop, Stream is not needed here

mikeoz0 replied:
done.

bsideup commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,139 @@
+        List<String> list = new ArrayList<>();
+        list.add(firstSecret);
+        Stream.of(remainingSecrets).forEach(s -> list.add(s));
+        if (secretsMap.containsKey(path)) {
```
`secretsMap.computeIfAbsent(path, ArrayList::new).addAll(list)`

mikeoz0 replied:
Couldn't quite get this exact command to work, but have reworked it and is cleaner now.

bsideup commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,139 @@
+public class VaultContainer<SELF extends VaultContainer<SELF>> extends GenericContainer<SELF>
+        implements LinkableContainer {
+
+    private boolean preloadSecret = false;
```
do we really need this flag? Can't we just assume that if secretsMap is not empty then we need to preload secrets? :)

mikeoz0 replied:
👍 removed it.

Commit:
Updated per review comments

updated per review comments

cleanup

mikeoswald commented:
@bsideup I've done my best to address your review comments. Can you take a look at the changes? Thanks

Commit:
updated javadoc

Commit:
updated test and Vault Container to ensure different ports can be passed

kiview commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,124 @@
+        setStartupAttempts(3);
+        withCreateContainerCmdModifier(cmd -> cmd.withCapAdd(IPC_LOCK));
+        if(!isVaultPortRequested()){
+            withEnv("VAULT_ADDR", "http://0.0.0.0:8200");
```
Port 8200 is used in multiple parts of the code, maybe have this in a constant?

kiview commented:
I've added some small comments, the rest looks good to me.

Commit:
Added constant for port 8200, and removed distributionManagement section
from pom for now

mikeoswald commented:
@kiview thanks for the review. I have updated the PR per your comments.

Commit:
Removed setPortBindings if port is not passed. This is for tests that
don't want to expose any ports and would rather have containers linked
in tests.

mikeoswald commented:
I just pushed one last change, I removed the setPortBindings if they don't specify a port to bind on. That way if people want to link to this container in tests they shouldn't need to expose ports. Just a cleaner approach. 

Thanks, I think this is ready for merge.

kiview commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,132 @@
+public class VaultContainer<SELF extends VaultContainer<SELF>> extends GenericContainer<SELF>
+        implements LinkableContainer {
+
+    private static final String VAULT_PORT="8200";
```
Please surround binary operators with two spaces (I assume it's a refactoring error here ;)).

kiview commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,132 @@
+        setStartupAttempts(3);
+        withCreateContainerCmdModifier(cmd -> cmd.withCapAdd(IPC_LOCK));
+        if(!isVaultPortRequested()){
+            withEnv("VAULT_ADDR", "http://0.0.0.0:"+VAULT_PORT);
```
Missing spaces here as well

kiview commented on src/main/java/org/testcontainers/vault/VaultContainer.java:
```diff
@@ -0,0 +1,132 @@
+        setVaultPortRequested(true);
+        String vaultPort = String.valueOf(port);
+        withEnv("VAULT_ADDR", "http://0.0.0.0:"+VAULT_PORT);
+        setPortBindings(Arrays.asList(vaultPort+":"+VAULT_PORT));
```
Missing spaces

kiview reviewed (changes requested):
Just a tiny code style change, sorry for nitpicking, I think this happened when extracting the constant. I will merge afterwards 👍

Commit:
Fixed spacing for formatting

Commit:
Fixed spacing for formatting
